File: benchmark/ToQUBO/tsp.jl
using CSV
using JuMP
using ToQUBO

function tsp(n::Int; clear_gc::Bool = false)
    clear_gc && GC.gc()

    # Problem Data
    D = fill(10, (n,n))

    # Build Model
    t₀ = @timed(begin
        model = Model(ToQUBO.Optimizer)

        @variable(model, x[1:n, 1:n], Bin, Symmetric)

        @constraint(model, [i in 1:n], sum(x[i,:]) == 1)
        @constraint(model, [j in 1:n], sum(x[:,j]) == 1)

        @objective(model, Min, sum(D[i,j] * x[i,k] * x[j, (k % n)+1] for i = 1:n , j = 1:n, k = 1:n))
    end).time

    clear_gc && GC.gc()

    # Compile Model
    t₁ = @timed(optimize!(model)).time

    # Convert to QUBO
    t₂ = @timed(Q, α, β = ToQUBO.qubo(unsafe_backend(model))).time

    return t₀, t₁, t₂
end

function main(initial_size::Int, max_size::Int, step::Int)
    results = []

    tsp(2) # avoid time-to-first-solve
    
    for n in initial_size:step:max_size
        t₀, t₁, t₂ = tsp(n; clear_gc=true)
        tₜ = t₀ + t₁ + t₂

        println(
            """
            -------------------------------------
            Variables: $(n * n)
            Model................ $(t₀)
            Compilation.......... $(t₁)
            Conversion........... $(t₂)
            Total elapsed time... $(tₜ)
            """
        )

        push!(results, (n * n, tₜ, t₀, t₁ + t₂))
    end

    csv_path = joinpath(@__DIR__, "results.csv")
    CSV.write(csv_path, sort(collect(results)), header = ["n_var","time","jump_time","toqubo_time"])
end

main(5,100,5)

File: benchmark/qubovert/tsp.py
import time
import pandas   as pd
import numpy    as np
import qubovert as qv
from pathlib import Path

__DIR__ = Path(__file__).parent

def tsp(n: int, lam:float = 5.0):
    # Problem Data
    D = 10.0 * np.ones((n, n), dtype=float)

    # Create Model
    t0 = time.time()
    
    model = 0.0

    # Add variables
    x = [[qv.boolean_var(f"x[{i},{j}]") for j in range(n)] for i in range(n)]

    # Add objective function
    for i in range(n):
        for j in range(n):
            for k in range(n):
                model += D[i][j] * x[k][i] * x[(k + 1) % n][j]
    
    # Add constraints
    for i in range(n):
        model.add_constraint_eq_zero(sum(x[i][j] for j in range(n)) - 1, lam=lam)

    for j in range(n):
        model.add_constraint_eq_zero(sum(x[i][j] for i in range(n)) - 1, lam=lam)

    # Convert to QUBO
    t1 = time.time()
    
    qubo = model.to_qubo()
    
    # Stop the count!
    t2 = time.time()

    return t1-t0, t2-t1

def main(init_size, max_size, step):
    results = {"nvar":[], "time":[]}

    for n in range(init_size, max_size+step, step):
        model_time, compiler_time, convert_time = tsp(n)
        total_time = model_time + compiler_time + convert_time

        print(
f"""\
-----------------------------
Variables: {n * n} ({n} sites)
Model................ {model_time:7.3f}
Compilation.......... {compiler_time:7.3f}
Conversion........... {convert_time:7.3f}
Total elapsed time... {total_time:7.3f}
""",
flush = True
        )

        results["nvar"].append(n * n)
        results["time"].append(total_time)

    df = pd.DataFrame(results)
    df.to_csv(__DIR__.joinpath("results.csv"), index = False)

if __name__ == '__main__':
    main(5, 100, 5)
